### ISSUE TITLE
CAdvisor stats collection should be configurable.

### ISSUE THREAD

**Issue by kannon92** [kubernetes/kubernetes#123340]
### What would you like to be added?

```
// TODO(vmarmol): Make configurable.
// The amount of time for which to keep stats in memory.
const statsCacheDuration = 2 * time.Minute
const maxHousekeepingInterval = 15 * time.Second
const defaultHousekeepingInterval = 10 * time.Second
const allowDynamicHousekeeping = true
```

These stats should be configurable in KubeletConfig. 

We should keep the defaults set if the user does not specify them in the KubeletConfig. 

### Why is this needed?

It would be nice to change these during e2e tests and I also think users would like the ability to change this.

**Comment by kannon92** [comment]
WDYT @haircommander ?

**Comment by haircommander** [comment]
hm considering we're moving away from cadvisor stats collection, I wonder if our effort would be better put into configuring the cri stats provider and giving users options to configure collection of those?

**Comment by dims** [comment]
> hm considering we're moving away from cadvisor stats collection, I wonder if our effort would be better put into configuring the cri stats provider and giving users options to configure collection of those?

+10000

**Comment by kannon92** [comment]
Thats true but we still don't have a story on disk collection so not sure if cadvisor is being dropped for a bit.

**Comment by haircommander** [comment]
ah that's a good point, some disk stats are remaining in cadvisor

**Comment by harshanarayana** [comment]
@kannon92  would be happy to take a look at this once this changes are approved.

**Comment by roobre** [comment]
I'd like to plusone this with my use case as well. I run Kubernetes on a couple clusters with a fair number of pods per node, around ~200. 

For the last few months, I've noticed that the kubelet process is by far the most CPU intensive on the whole node, with CPU spikes of +200% CPU, into what seems to be a periodic pattern:

![image](https://github.com/user-attachments/assets/1e06e3c8-7495-4fe3-af6e-4c49320899e0)

Today I finally got some motivation to enable profiling and show what is the kubelet is doing exactly, which turned out to be `ncdu` (or the cadvisor version of it):

![profile002](https://github.com/user-attachments/assets/92b60ce1-5bb9-4aaf-9ac4-052de13c0ac2)

Kubelet/cadvisor spent 76 out of 60s of the profiling time in `realFsHandler.trackUsage`, basically walking directories and issuing syscalls to get its size.

I'd love to be able to tweak cadvisor settings to tune this to my workloads, e.g. reducing the FS scan period from one minute https://github.com/kubernetes/kubernetes/blob/a87612b6676723b34a5b3d2d80ab4e04552221ae/vendor/github.com/google/cadvisor/container/common/fsHandler.go#L57 to something like 10, or if this were not possible, perhaps disabling disk usage monitoring entirely.

**Comment by kannon92** [comment]
@roobre thank you for posting your findings.

Would you be interested in contributing a feature for this?

**Comment by roobre** [comment]
Hey @kannon92, I can give it a stab, provided some direction on how you'd like the feature to look like. My initial idea was to have a `cadvisorConfig` entry in the kubelet config object, and inside it add fields for these constants and enabling certain collectors:

```yaml
enableServer: true
# ...
cadvisor:
  statsCacheDuration:
  maxHousekeepingInterval:
  defaultHousekeepingInterval:
  allowDynamicHousekeeping:
  fsScanPeriod:
  includedMetrics:
    cpu: true
    diskUsaged: false
```

If this looks good I can try and see how hard it would be to propagate this struct down the call stack.

**Comment by ffromani** [comment]
> Hey @kannon92, I can give it a stab, provided some direction on how you'd like the feature to look like. My initial idea was to have a `cadvisorConfig` entry in the kubelet config object, and inside it add fields for these constants and enabling certain collectors:
> 
> ```yaml
> enableServer: true
> # ...
> cadvisor:
>   statsCacheDuration:
>   maxHousekeepingInterval:
>   defaultHousekeepingInterval:
>   allowDynamicHousekeeping:
>   fsScanPeriod:
>   includedMetrics:
>     cpu: true
>     diskUsaged: false
> ```
> 
> If this looks good I can try and see how hard it would be to propagate this struct down the call stack.

(I'm in favour and interested in this change)
please be aware that `kubeletconfig` changes will require API review, which can take some time, and user-facing changes MAY require a KEP process.
Long story short, user-facing changes are hardly served by a simple PR, it is likely they will need more process. I recommend to check before to get started.

**Comment by roobre** [comment]
@ffromani That sounds fair. Where can I read more about the steps required to get something like this done?

**Comment by kannon92** [comment]
I think this should require a KEP but it shouldn’t be too complicated. It will make it easier for tracking the changes. 

We closed features for 1.31 so you could play around with an implementation and work on the KEP for 1.32.

**Comment by dims** [comment]
@kannon92 also remember we *are* trying to get rid of cadvisor too!

**Comment by ffromani** [comment]
> @ffromani That sounds fair. Where can I read more about the steps required to get something like this done?

this is a good intro to the process: https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/0000-kep-process/README.md

you can find examples of the KEP process looking at the merged KEPs: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node

I'd recommend to start socializing the proposal on sig-node slack channel and/or mailing list or attending the sig-node weekly meetings (it's on tuesday). Getting input here should help a lot in finding the best way forward and save time.

Also, thanks @dims for the reminder! The issue reported here is real and a fix would improve the user experience, but considering that we want to (slowly?) phase out cadvisor, we should probably do minimal changes. Here the input of the sig-node community is highly beneficial.

**Comment by kannon92** [comment]
> @kannon92 also remember we _are_ trying to get rid of cadvisor too!

Yes. But we need to GA CRI stats first which seems to be not happening in the near future. And then we still use cadvisor for disk stats. Disk stats were out of scope for CRI stats so cadvisor will still be around.

**Comment by kannon92** [comment]
Xref: https://github.com/kubernetes/kubernetes/pull/126052

**Comment by haircommander** [comment]
Personally I would prefer we focus our energy on getting CRI stats feature moving forward (currently waiting on a containerd implementation of the metrics collection). I think adding kubelet configuration fields for a component we're actively trying to remove from the stats collection process is counter intuitive, especially with how long APIs survive in kubernetes.

**Comment by haircommander** [comment]
having configurable disk collection for cadvisor from the kubelet config would work for me as well, if that is a feature that is of interest to an end user

**Comment by hsunwenfang** [comment]
> Personally I would prefer we focus our energy on getting CRI stats feature moving forward (currently waiting on a containerd implementation of the metrics collection). I think adding kubelet configuration fields for a component we're actively trying to remove from the stats collection process is counter intuitive, especially with how long APIs survive in kubernetes.

Hi @haircommander 
Just bumped into this
When you say "getting CRI stats feature moving forward"
Where are these features located now?
Is it here?
https://github.com/kubernetes/cri-api/tree/master
Is this even the correct repo?
Or is there a specific folder for it?
thx!

**Comment by haircommander** [comment]
it's largely in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go and the CRI-API as you found

**Comment by kannon92** [comment]
cc @yangyefei90

**Comment by kannon92** [comment]
@haircommander it seems that every few weeks I find more stats that we rely on cadvisor for.

CRIStats Provider will provide stats for cadvisor cgroups for containers/pods.

CAdvisor will still be in use for hardware/disk/node level stats.

So it seems that this issue would be relevant as we will not be able to drop cadvisor anytime soon.

**Comment by SuperQ** [comment]
One drive-by performance comment (CC @roobre)

One thing you may want to consider is adjusting the [`GOMAXPROCS`](https://pkg.go.dev/runtime#hdr-Environment_Variables) environment variable for the kubelet itself. If you have nodes with a number of CPUs this can help reduce the peaks. For exampke, you can use `GOMAXPROCS=2` to limit the kubelet to only use 2 CPUs. Of course, this could make the use sustained over a longer period of time. But I've seen some overall performance improvements on larger systems (32+ CPUs) due to the reduced CPU cache blowout where Go tries to run a lot of goroutines in parallel.

**Comment by roobre** [comment]
Hey @SuperQ, I appreciate the suggestion! I might as well try this to see what happens. I'm slightly concerned on whether that can have an impact on the time it takes for the kubelet to do other things, but seems worth trying anyway.

**Comment by raywainman** [comment]
Hey team,

Curious where things are at with this feature request?

I'm approaching this from the autoscaling angle (via SIG Autoscaling) where we'd like to get fresher metrics to make faster autoscaling decisions. Being able to customize the stats collection interval here is something that is very interesting for us.

It looks like this was initially set to 10s in #19374 nearly 10 years ago (!) - the demands of autoscaling have evolved significantly since then :)

Happy to help out here and contribute however I can.

Thanks!

**Comment by NahumLitvin** [comment]
We hit significant kubelet CPU from cAdvisor ProcessMetrics (/proc scan per thread). On 3 prod clusters (~1000 nodes, 100+ pods/node) kubelet used 2–4 cores; disabling ProcessMetrics dropped to <0.1 cores (≈99.4% reduction).

**Proposal:**
- New alpha feature gate: `ConfigurableCAdvisorMetrics` (target 1.33)
- New `KubeletConfiguration.cadvisor.includedMetrics.processMetrics` (default `true`)
- Allows disabling ProcessMetrics while keeping defaults unchanged for compatibility

**Implementation branch:** https://github.com/NahumLitvin/kubernetes/tree/feature/configurable-cadvisor-metrics
**KEP draft:** 5000 (configurable cAdvisor metrics)
**Tests:** `go test ./pkg/kubelet/apis/config/...` (darwin); `go test ./pkg/kubelet/cadvisor` (linux/amd64, go1.25)

Guidance on next steps (KEP + impl PR) appreciated.

**Comment by kannon92** [comment]
As I posted this two years ago now, it does seem that CRI Stats is further along.

I'm not as sure we should go forward with this honestly.

**Comment by kannon92** [comment]
> Guidance on next steps (KEP + impl PR) appreciated.

It would be worth going to a sig-node meeting to convey your findings.

The KEP process is defined in kubernetes/enhancements.

**Comment by kannon92** [comment]
> We hit significant kubelet CPU from cAdvisor ProcessMetrics (/proc scan per thread). On 3 prod clusters (~1000 nodes, 100+ pods/node) kubelet used 2–4 cores; disabling ProcessMetrics dropped to <0.1 cores (≈99.4% reduction).
> 
> **Proposal:**
> 
> * New alpha feature gate: `ConfigurableCAdvisorMetrics` (target 1.33)
> * New `KubeletConfiguration.cadvisor.includedMetrics.processMetrics` (default `true`)
> * Allows disabling ProcessMetrics while keeping defaults unchanged for compatibility
> 
> **Implementation branch:** https://github.com/NahumLitvin/kubernetes/tree/feature/configurable-cadvisor-metrics **KEP draft:** 5000 (configurable cAdvisor metrics) **Tests:** `go test ./pkg/kubelet/apis/config/...` (darwin); `go test ./pkg/kubelet/cadvisor` (linux/amd64, go1.25)
> 
> Guidance on next steps (KEP + impl PR) appreciated.

Disabling ProcessMetrics essentially means that Kubelet will no longer monitor process pid usage which does become a breaking change for eviction manager.

**Comment by kannon92** [comment]
@NahumLitvin

https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/kep.yaml

CRI Stats is now beta in 1.35 (I think it may be beta off).

With your feature you are looking at 1+ year of getting this into production. 

Have you tried enabling CRI Stats and seeing what items you could propose to fix your use case?

New KEPs take a lot longer than existing KEPs.

**Comment by NahumLitvin** [comment]
Thanks for the quick follow-up. A few points:

- CRI stats KEP (2371) is promising, but on managed platforms (EKS, and per most platform engineers I know) 1.35 is years away. We need a stopgap now to stop burning 2–4 cores per kubelet.
- This feature gate is alpha and **off by default**; defaults stay the same. It’s opt-in for operators hit by ProcessMetrics cost.
- Eviction isn’t impacted: PID pressure uses `/proc/loadavg` (`pidlimit_linux.go`), not cAdvisor ProcessMetrics.
- Others have hit the same overhead: https://github.com/kubernetes/kubernetes/issues/99183 and https://github.com/kubernetes/kubernetes/issues/82112.
- We already patch kubelet in our AMI to disable ProcessMetrics to keep clusters stable.

**Comment by NahumLitvin** [comment]
Understood that EKS won’t ship this until it’s beta/on-by-default. We still need an upstream toggle so self-managed clusters (and folks already patching kubelet) can stop burning 2–4 cores right now. Gate stays alpha/off by default; defaults unchanged.

If there’s a path to give current operators relief without blocking on managed distro timelines, I’m all for it. Could we keep this open and at least get `/ok-to-test` so we can validate?
